packages/expo-auth-session/src/providers/Google.ts

```diff
@@ -0,0 +1,225 @@
+import { Platform } from 'react-native';
+
+import { useAuthRequestResult, useLoadedAuthRequest } from '../AuthRequestHooks';
+import {
+  AuthRequest,
+  AuthRequestConfig,
+  AuthRequestPromptOptions,
+  AuthSessionRedirectUriOptions,
+  AuthSessionResult,
+  DiscoveryDocument,
+  generateHexStringAsync,
+  makeRedirectUri,
+  Prompt,
+  ResponseType,
+} from '../AuthSession';
+import { ProviderAuthRequestConfig } from './Provider.types';
+
+const settings = {
+  windowFeatures: { width: 515, height: 680 },
+  minimumScopes: [
+    'openid',
+    'https://www.googleapis.com/auth/userinfo.profile',
+    'https://www.googleapis.com/auth/userinfo.email',
+  ],
+};
+
+export const discovery: DiscoveryDocument = {
+  authorizationEndpoint: 'https://accounts.google.com/o/oauth2/v2/auth',
+  tokenEndpoint: 'https://oauth2.googleapis.com/token',
+  revocationEndpoint: 'https://oauth2.googleapis.com/revoke',
+  userInfoEndpoint: 'https://openidconnect.googleapis.com/v1/userinfo',
+};
+
+export interface GoogleAuthRequestConfig extends ProviderAuthRequestConfig {
+  /**
+   * If the user's email address is known ahead of time, it can be supplied to be the default option.
+   * If the user has approved access for this app in the past then auth may return without any further interaction.
+   */
+  loginHint?: string;
+
+  /**
+   * Proxy client ID for use in the Expo client on iOS and Android.
+   *
+   * This Google Client ID must be setup as follows:
+   *
+   * - **Application Type**: Web Application
+   * - **URIs**: https://auth.expo.io
+   * - **Authorized redirect URIs**: https://auth.expo.io/@your-username/your-project-slug
+   */
+  expoClientId?: string;
+
+  /**
+   * Expo web client ID for use in the browser.
+   *
+   * This Google Client ID must be setup as follows:
+   *
+   * - **Application Type**: Web Application
+   * - **URIs**: https://yourwebsite.com | https://localhost:19006
+   * - **Authorized redirect URIs**: https://yourwebsite.com | https://localhost:19006
+   */
+  webClientId?: string;
+
+  /**
+   * iOS native client ID for use in standalone, bare-workflow, and custom clients.
+   *
+   * This Google Client ID must be setup as follows:
+   *
+   * - **Application Type**: iOS Application
+   * - **Bundle ID**: Must match the value of `ios.bundleIdentifier` in your `app.json`.
+   * - Your app needs to conform to the URI scheme matching your bundle identifier.
+   *   - **Standalone**: Automatically added, do nothing.
+   *   - **Bare-workflow**: Run `npx uri-scheme add <your bundle id> --ios`
+   */
+  iosClientId?: string;
+  /**
+   * Android native client ID for use in standalone, bare-workflow, and custom clients.
+   *
+   * This Google Client ID must be setup as follows:
+   *
+   * - **Application Type**: Android Application
+   * - **Package name**: Must match the value of `android.package` in your `app.json`.
+   * - **Signing-certificate fingerprint**: Run `openssl rand -base64 32 | openssl sha1 -c` for the results.
```

Comment:
Probably need something like `expo fetch:android:hashes` instead.

Reply:
you may already know this and i'm just misreading the comment but `expo fetch:android:hashes` does this already, the value you're looking for is "Google Certificate Fingerprint"

Reply:
I mean if you run it in any Expo project that isn't built for prod you get:
```
Configuring credentials for bacon in project native-component-list
There is no valid Keystore defined for this app
```
Effectively meaning that in dev you need to set it to one value then change it in prod or create another Google key which is worse.

Reply:
The current way everything is set-up, íf you want to use Google App Signing AND you want expo builds to have a different key store, you _must_ have two google keys. Google's "recent" restrictions on those keys are weird, as almost all google systems allow for multiple values (all other android key restrictions). 

You might not be aware, but if you read through https://github.com/expo/expo/issues/9391, you will need **3 separate keys** anyway.

Reply:
I've just opened an issue to make the CLI return a keystore on demand without requiring a build https://github.com/expo/expo-cli/issues/2456 this would effectively cut out one more key upfront for Android auth.

Reply:
That's a good idea. I don't think having to have 3-4 keys is that problematic, but the fact that the "openssl" command is cryptic as best makes this a very good win, I think.

Reply:
I think it's problematic currently because currently you:

- get fake hash -> register Google key with fake hash -> `expo build` it -> get actual key -> register new Google key -> update codebase -> `expo publish` again.

After this you could

- get future hash -> register Google key with fake hash -> `expo build` it.

Much easier to document, test, and develop.

Reply:
Yeah, I think making this process for Google Play Store simpler is much much better. I did exactly what you say in the "before" list, and it's confusing if you don't know what's going on.

But that's not what I meant with it not being problematic. I meant that having a key for expo-non-google builds and google-signed builds _does_ make sense -- so has nothing to do with the process of acquiring that key, but with the fact that we still should be able to support at least **2 keys** for Android projects, *1 key* for iOS and *1 key* for Web.

Why?

If there can only be one Android key, non-google signed builds will no longer work, making the internal testing process far worse, and also make it much harder to have a google signed build for the play store and a expo signed build for all the other stores (amazon, samsung etc.).